### DISCUSSION TITLE
Synchronizing Project Updates from Main Repo

### DISCUSSION THREAD

**Discussion by jamestouri** [medplum/medplum#1968] [Ideas]
As projects to showcase our code and solutions become more mission critical working with customers, we need to maintain a high standard of quality with every project we show.  Right now, we have _a lot_ of projects.

We can create a workflow in the github actions repo that will be able to change the 

The high level overview of this will essentially involve creating a workflow file in `.github/workflows` of the Medplum repo. And running that will run through changes in file paths and identifying if any of the projects in the main repo has been changed. If it has, the `[actions/github-script](https://github.com/actions/github-script)` will call an API request to update the listening repositories. The listening repo will take a webhook payload and find out the changes to make, before creating a pull request. 

The breakdown of the steps will be as follows:
- Create a new file in the `.github/workflows` directory. The file will define the workflow that will run when a pull request is merged into the main repo. 
- The workflow file will have an action that will make API requests based on the path the code changes.
 Call the [actions/checkout](https://github.com/actions/checkout) action to see the code changes 
- Using [actions/github-script](https://github.com/actions/github-script), we make an API request to the listening repo and create a PR for the code changes

A few characteristics we will have in relation from an example project to the main repo that we need to think about:
- Every project is self-contained: it will have their own package.json, node_modules, etc. 
- When running `npm i` or `npm run build` from the medplum root directory, if should build the example projects as well 
- The package.json file in the subdirectories should be synced up with the dependencies that the root package.json file has. 
  - This could be done manually via script

**Comment by rahul1** [comment]
@jamestouri this seems like a reasonable high-level plan. If I understand correctly, does this mean that every commit will be sync'ed to the target repo? 

Also will the package structure look like for example apps stored in our main repo? Will they all be nested in a single package, or will they be flat

**Reply by jamestouri** [reply]
@rahul1 every time that a branch merges to main, the github actions script will check if there were any code changes in an `examples/*` file. 

> Also will the package structure look like for example apps stored in our main repo? Will they all be nested in a single package, or will they be flat

They should be flat. Ex. there's a folder `examples/hello-world` and there are 3 files of code in it, the repo will be called hello-world and it'll have the 3 files of code in the root.

**Comment by jamestouri** [comment]
Updates: 
- Have a PR out for this https://github.com/medplum/medplum/pull/1979 

Next Steps: 
- Move medplum-hello-world and prepare it to be synced to the listening repo 
- Adding the examples/* to the npm workspaces as we will start building the node_modules along with the code in packages/*
- shell script that updates nested dependencies if there is a change in devDependency versions in root package.json file

**Comment by codyebberson** [comment]
Done!  Thanks @jamestouri!
